<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Indexing &mdash; xftsim 0.0.1-dev9 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Data structures" href="struct.html" />
    <link rel="prev" title="Nuts and bolts" href="nutsandbolts.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/xftsimlogomediumwhite.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../gettingstarted/getting_started.html">Getting started</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="user_guide.html">User guide</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="nutsandbolts.html">Nuts and bolts</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Indexing</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Component-indexing">Component indexing</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Variant-indexing">Variant indexing</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Sample-indexing">Sample indexing</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="struct.html">Data structures</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="simulation.html">Anatomy of a simulation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">xftsim</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="user_guide.html">User guide</a></li>
          <li class="breadcrumb-item"><a href="nutsandbolts.html">Nuts and bolts</a></li>
      <li class="breadcrumb-item active">Indexing</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/userguide/indexing.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Indexing">
<h1>Indexing<a class="headerlink" href="#Indexing" title="Permalink to this heading"></a></h1>
<p>In the course of constructing simulations, we very frequently want to refer to specific individuals, specific variants, or specific phenotyping components. For example, when we create a haplotype array, rows will correspond in individuals and columns will correspond to haploid sites.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">xftsim.index</span></code> submodule implements classes corresponding to these specific cases:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">xft.index.SampleIndex</span></code> indexes individuals</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">xft.index.HaploidVariantIndex</span></code> and <code class="docutils literal notranslate"><span class="pre">xft.index.DiploidVariantIndex</span></code> index genetic variants</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">xft.index.ComponentIndex</span></code> indexes phenotypes and components of phenotypes</p></li>
</ul>
<p>All the above are instances of the <code class="docutils literal notranslate"><span class="pre">xft.index.XftIndex</span></code> superclass and can be represented as Pandas DataFrames using the <code class="docutils literal notranslate"><span class="pre">.frame</span></code> property:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">xftsim</span> <span class="k">as</span> <span class="nn">xft</span>

<span class="c1">## Here&#39;s a ComponentIndex instance. Ignore the details for now.</span>
<span class="n">cindex</span> <span class="o">=</span> <span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">ComponentIndex</span><span class="o">.</span><span class="n">from_product</span><span class="p">((</span><span class="s1">&#39;phenotype_1&#39;</span><span class="p">,</span> <span class="s1">&#39;phenotype_2&#39;</span><span class="p">),</span>
                                               <span class="p">(</span><span class="s1">&#39;additiveGenetic&#39;</span><span class="p">,</span> <span class="s1">&#39;additiveNoise&#39;</span><span class="p">,</span> <span class="s1">&#39;phenotype&#39;</span><span class="p">))</span>

<span class="c1">## We can turn this in a Pandas data.frame</span>
<span class="n">cframe</span> <span class="o">=</span> <span class="n">cindex</span><span class="o">.</span><span class="n">frame</span>
<span class="n">cframe</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>phenotype_name</th>
      <th>component_name</th>
      <th>vorigin_relative</th>
    </tr>
    <tr>
      <th>component</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>phenotype_1.additiveGenetic.proband</th>
      <td>phenotype_1</td>
      <td>additiveGenetic</td>
      <td>-1</td>
    </tr>
    <tr>
      <th>phenotype_1.additiveNoise.proband</th>
      <td>phenotype_1</td>
      <td>additiveNoise</td>
      <td>-1</td>
    </tr>
    <tr>
      <th>phenotype_1.phenotype.proband</th>
      <td>phenotype_1</td>
      <td>phenotype</td>
      <td>-1</td>
    </tr>
    <tr>
      <th>phenotype_2.additiveGenetic.proband</th>
      <td>phenotype_2</td>
      <td>additiveGenetic</td>
      <td>-1</td>
    </tr>
    <tr>
      <th>phenotype_2.additiveNoise.proband</th>
      <td>phenotype_2</td>
      <td>additiveNoise</td>
      <td>-1</td>
    </tr>
    <tr>
      <th>phenotype_2.phenotype.proband</th>
      <td>phenotype_2</td>
      <td>phenotype</td>
      <td>-1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<section id="Component-indexing">
<h2>Component indexing<a class="headerlink" href="#Component-indexing" title="Permalink to this heading"></a></h2>
<section id="Overview">
<h3>Overview<a class="headerlink" href="#Overview" title="Permalink to this heading"></a></h3>
<p>The most important indexer to undeestand is the <code class="docutils literal notranslate"><span class="pre">xft.index.ComponentIndex</span></code>. In <code class="docutils literal notranslate"><span class="pre">xftsim</span></code>, individual level data with respect to a given sample (the “proband”) is indexed as follows:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">phenotype_name</span></code>, phenotypes,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">component_name</span></code>, components of phenotypes,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">vorigin_relative</span></code>, relationship of originator of component to proband (see below).</p></li>
</ul>
<p>Components are essentially sub-phenotypes and include the phenotype itself. For example, euppose we assume that the phenotype height is comprised of two components: an additive genetic component and an additive, individual-specific noise component. The corresponding component index looks like this:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">ComponentIndex</span><span class="o">.</span><span class="n">from_product</span><span class="p">(</span><span class="n">phenotype_name</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;height&#39;</span><span class="p">),</span>
                                      <span class="n">component_name</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;additiveGenetic&#39;</span><span class="p">,</span> <span class="s1">&#39;additiveNoise&#39;</span><span class="p">,</span> <span class="s1">&#39;phenotype&#39;</span><span class="p">),</span>
                                      <span class="n">vorigin_relative</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ComponentIndex&gt;
  3 components of 1 phenotype spanning 1 generation
                               phenotype_name   component_name  \
component
height.additiveGenetic.proband         height  additiveGenetic
height.additiveNoise.proband           height    additiveNoise
height.phenotype.proband               height        phenotype

                                vorigin_relative
component
height.additiveGenetic.proband                -1
height.additiveNoise.proband                  -1
height.phenotype.proband                      -1
</pre></div></div>
</div>
<p>Note that we have added a third component <code class="docutils literal notranslate"><span class="pre">phenotype</span></code> to represent the sums of the first two components. Components can be as general or as specific as we like.</p>
<p>In many cases, we can ignore <code class="docutils literal notranslate"><span class="pre">vorigin_relative</span></code>, but it’s essential if we want to model intergenerational phenotypic effects such as vertical transmission. It works like this:</p>
<p><code class="docutils literal notranslate"><span class="pre">vorigin_relative</span></code> is a binary encoding that represents relationship to a proband as follows:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p><code class="docutils literal notranslate"><span class="pre">vorigin_relative</span></code></p></th>
<th class="head"><p>relationship to proband</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>-1</p></td>
<td><p>self</p></td>
</tr>
<tr class="row-odd"><td><p>0</p></td>
<td><p>mother</p></td>
</tr>
<tr class="row-even"><td><p>1</p></td>
<td><p>father</p></td>
</tr>
<tr class="row-odd"><td><p>00</p></td>
<td><p>maternal grandmother</p></td>
</tr>
<tr class="row-even"><td><p>10</p></td>
<td><p>maternal grandfather</p></td>
</tr>
<tr class="row-odd"><td><p>01</p></td>
<td><p>paternal grandmother</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p>paternal grandfather</p></td>
</tr>
<tr class="row-odd"><td><p>000</p></td>
<td><p>maternal grandmother’s mother</p></td>
</tr>
<tr class="row-even"><td><p>100</p></td>
<td><p>maternal grandmother’s father</p></td>
</tr>
<tr class="row-odd"><td><p>…</p></td>
<td><p>…</p></td>
</tr>
</tbody>
</table>
<p>and so forth. In most cases, considering maternal or paternal effects is sufficent. Suppose we’d like to jointly analyze trio’s heights, bone mineral density (BMD), as well as the heritable and nonheritable components thereof. In this case, our component index would look like this:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">ComponentIndex</span><span class="o">.</span><span class="n">from_product</span><span class="p">(</span><span class="n">phenotype_name</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;height&#39;</span><span class="p">,</span><span class="s1">&#39;BMD&#39;</span><span class="p">),</span>
                                      <span class="n">component_name</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;genetic&#39;</span><span class="p">,</span> <span class="s1">&#39;noise&#39;</span><span class="p">,</span> <span class="s1">&#39;phenotype&#39;</span><span class="p">),</span>
                                      <span class="n">vorigin_relative</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ComponentIndex&gt;
  3 components of 2 phenotypes spanning 2 generations
                         phenotype_name component_name  vorigin_relative
component
height.genetic.proband           height        genetic                -1
height.genetic.mother            height        genetic                 0
height.genetic.father            height        genetic                 1
height.noise.proband             height          noise                -1
height.noise.mother              height          noise                 0
height.noise.father              height          noise                 1
height.phenotype.proband         height      phenotype                -1
height.phenotype.mother          height      phenotype                 0
height.phenotype.father          height      phenotype                 1
BMD.genetic.proband                 BMD        genetic                -1
BMD.genetic.mother                  BMD        genetic                 0
BMD.genetic.father                  BMD        genetic                 1
BMD.noise.proband                   BMD          noise                -1
BMD.noise.mother                    BMD          noise                 0
BMD.noise.father                    BMD          noise                 1
BMD.phenotype.proband               BMD      phenotype                -1
BMD.phenotype.mother                BMD      phenotype                 0
BMD.phenotype.father                BMD      phenotype                 1
</pre></div></div>
</div>
<p>There are several ways to construct a component index, the first of which we’ve used several times already:</p>
</section>
<section id="Constructing-a-component-index-from-a-cartestian-product">
<h3>Constructing a component index from a cartestian product<a class="headerlink" href="#Constructing-a-component-index-from-a-cartestian-product" title="Permalink to this heading"></a></h3>
<p>To generate a component index by expanding combinations of <code class="docutils literal notranslate"><span class="pre">phenotype_name</span></code>, <code class="docutils literal notranslate"><span class="pre">component_name</span></code>, and <code class="docutils literal notranslate"><span class="pre">vorigin_relative</span></code>, we can use the <code class="docutils literal notranslate"><span class="pre">.from_product()</span></code> method:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">ComponentIndex</span><span class="o">.</span><span class="n">from_product</span><span class="p">(</span><span class="n">phenotype_name</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;height&#39;</span><span class="p">,),</span>
                                      <span class="n">component_name</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;phenotype&#39;</span><span class="p">),</span>
                                      <span class="n">vorigin_relative</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ComponentIndex&gt;
  1 component of 1 phenotype spanning 2 generations
                         phenotype_name component_name  vorigin_relative
component
height.phenotype.proband         height      phenotype                -1
height.phenotype.father          height      phenotype                 1
</pre></div></div>
</div>
<p>See above for further examples.</p>
</section>
<section id="Constructing-a-component-index-for-specific-components">
<h3>Constructing a component index for specific components<a class="headerlink" href="#Constructing-a-component-index-for-specific-components" title="Permalink to this heading"></a></h3>
<p>We can also manually specify arguments:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cindex</span> <span class="o">=</span> <span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">ComponentIndex</span><span class="p">(</span><span class="n">phenotype_name</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;height&#39;</span><span class="p">,</span><span class="s1">&#39;BMD&#39;</span><span class="p">),</span>
                                  <span class="n">component_name</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;phenotype&#39;</span><span class="p">,</span><span class="s1">&#39;genetic&#39;</span><span class="p">),</span>
                                  <span class="n">vorigin_relative</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="n">cindex</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ComponentIndex&gt;
  2 components of 2 phenotypes spanning 2 generations
                         phenotype_name component_name  vorigin_relative
component
height.phenotype.proband         height      phenotype                -1
BMD.genetic.father                  BMD        genetic                 1
</pre></div></div>
</div>
<p>or provide a Pandas DataFrame with the same information:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">ComponentIndex</span><span class="p">(</span><span class="n">frame</span> <span class="o">=</span> <span class="n">cindex</span><span class="o">.</span><span class="n">frame</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ComponentIndex&gt;
  2 components of 2 phenotypes spanning 2 generations
                         phenotype_name component_name  vorigin_relative
component
height.phenotype.proband         height      phenotype                -1
BMD.genetic.father                  BMD        genetic                 1
</pre></div></div>
</div>
</section>
<section id="Constructing-a-generic-component-index">
<h3>Constructing a generic component index<a class="headerlink" href="#Constructing-a-generic-component-index" title="Permalink to this heading"></a></h3>
<p>Finally, we can construct a generic component index by providing the number of phenotypes, <code class="docutils literal notranslate"><span class="pre">k_total</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">ComponentIndex</span><span class="p">(</span><span class="n">k_total</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ComponentIndex&gt;
  1 component of 3 phenotypes spanning 1 generation
                  phenotype_name component_name  vorigin_relative
component
0.generic.proband              0        generic                -1
1.generic.proband              1        generic                -1
2.generic.proband              2        generic                -1
</pre></div></div>
</div>
</section>
</section>
<section id="Variant-indexing">
<h2>Variant indexing<a class="headerlink" href="#Variant-indexing" title="Permalink to this heading"></a></h2>
<p>Genetic data can be influenced with either <code class="docutils literal notranslate"><span class="pre">xft.index.HaploidVariantIndex</span></code> or <code class="docutils literal notranslate"><span class="pre">xft.index.DiploidVariantIndex</span></code> objects. The former, which indexes haploid sites, is used when distinguishing between homologous sites is necessary, such as during meioisis. It is trivial to switch between these indices so we will focus our introduction on <code class="docutils literal notranslate"><span class="pre">xft.index.DiploidVariantIndex</span></code>.</p>
<p>A <code class="docutils literal notranslate"><span class="pre">DiploidVariantIndex</span></code> tracks the following variant-level information</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">vid</span></code>, a vector of variant IDs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">chrom</span></code>, a vector of chromosome IDs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">zero_allele</span></code>, the allele corresponding to zeros</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">one_allele</span></code>, a vector of chromosome IDs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">af</span></code>, a vector of ancestral allele frequencies.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">annotation_array</span></code>, additional variant level annotations</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">h_copy</span></code>, haplotype copy (always ‘d’ for diploid data, ‘0’ or ‘1’ for haploid data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pos_bp</span></code>, physical position measured in basepairs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pos_cM</span></code>, position as measured in centiMorgans</p></li>
</ul>
<p>There are several ways to construct variant indices, though you’ll rarely need to do this in practice. Typically, you’ll use an automatically generated variant index, regardless of whether you’re using real or synthethic founder data.</p>
<p>### Constructing a variant index</p>
<p>To construct a generic variant index, you only need to provide the number of diploid variants <code class="docutils literal notranslate"><span class="pre">m</span></code> and the number of chromosomes <code class="docutils literal notranslate"><span class="pre">n_chrom</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vind</span> <span class="o">=</span> <span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">DiploidVariantIndex</span><span class="p">(</span><span class="n">m</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span><span class="n">n_chrom</span><span class="o">=</span><span class="mi">22</span><span class="p">)</span>
<span class="n">vind</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;DiploidVariantIndex&gt;
  500 diploid variants on 22 chromosome(s)
  MAF ranges from nan to nan
  0 annotation(s)
         vid  chrom zero_allele one_allele  af hcopy  pos_bp  pos_cM
variant
0.d        0      0           A          G NaN     d     NaN     NaN
1.d        1      0           A          G NaN     d     NaN     NaN
2.d        2      0           A          G NaN     d     NaN     NaN
3.d        3      0           A          G NaN     d     NaN     NaN
4.d        4      0           A          G NaN     d     NaN     NaN
...      ...    ...         ...        ...  ..   ...     ...     ...
495.d    495     21           A          G NaN     d     NaN     NaN
496.d    496     21           A          G NaN     d     NaN     NaN
497.d    497     21           A          G NaN     d     NaN     NaN
498.d    498     21           A          G NaN     d     NaN     NaN
499.d    499     21           A          G NaN     d     NaN     NaN

[500 rows x 8 columns]
</pre></div></div>
</div>
<p>Alternatively we can supply the above arguments individually (only <code class="docutils literal notranslate"><span class="pre">vid</span></code> is strictly necessary) or via a pandas DataFrame:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">DiploidVariantIndex</span><span class="p">(</span><span class="n">vid</span><span class="o">=</span><span class="n">vind</span><span class="o">.</span><span class="n">vid</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;DiploidVariantIndex&gt;
  500 diploid variants on 1 chromosome(s)
  MAF ranges from nan to nan
  0 annotation(s)
         vid  chrom zero_allele one_allele  af hcopy  pos_bp  pos_cM
variant
0.d        0      0           A          G NaN     d     NaN     NaN
1.d        1      0           A          G NaN     d     NaN     NaN
2.d        2      0           A          G NaN     d     NaN     NaN
3.d        3      0           A          G NaN     d     NaN     NaN
4.d        4      0           A          G NaN     d     NaN     NaN
...      ...    ...         ...        ...  ..   ...     ...     ...
495.d    495      0           A          G NaN     d     NaN     NaN
496.d    496      0           A          G NaN     d     NaN     NaN
497.d    497      0           A          G NaN     d     NaN     NaN
498.d    498      0           A          G NaN     d     NaN     NaN
499.d    499      0           A          G NaN     d     NaN     NaN

[500 rows x 8 columns]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">DiploidVariantIndex</span><span class="p">(</span><span class="n">frame</span><span class="o">=</span><span class="n">vind</span><span class="o">.</span><span class="n">frame</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;DiploidVariantIndex&gt;
  500 diploid variants on 22 chromosome(s)
  MAF ranges from nan to nan
  0 annotation(s)
         vid  chrom zero_allele one_allele  af hcopy  pos_bp  pos_cM
variant
0.d        0      0           A          G NaN     d     NaN     NaN
1.d        1      0           A          G NaN     d     NaN     NaN
2.d        2      0           A          G NaN     d     NaN     NaN
3.d        3      0           A          G NaN     d     NaN     NaN
4.d        4      0           A          G NaN     d     NaN     NaN
...      ...    ...         ...        ...  ..   ...     ...     ...
495.d    495     21           A          G NaN     d     NaN     NaN
496.d    496     21           A          G NaN     d     NaN     NaN
497.d    497     21           A          G NaN     d     NaN     NaN
498.d    498     21           A          G NaN     d     NaN     NaN
499.d    499     21           A          G NaN     d     NaN     NaN

[500 rows x 8 columns]
</pre></div></div>
</div>
<section id="Switching-between-haploid-and-diploid-indices">
<h3>Switching between haploid and diploid indices<a class="headerlink" href="#Switching-between-haploid-and-diploid-indices" title="Permalink to this heading"></a></h3>
<p>A <code class="docutils literal notranslate"><span class="pre">DiploidVariantIndex</span></code> can be converted to a haploid index and back via the <code class="docutils literal notranslate"><span class="pre">xft.index.DiploidVariantIndex.to_haploid()</span></code> and <code class="docutils literal notranslate"><span class="pre">xft.index.HaploidVariantIndex.to_diploid()</span></code> methods respectively:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hvind</span> <span class="o">=</span> <span class="n">vind</span><span class="o">.</span><span class="n">to_haploid</span><span class="p">()</span>
<span class="n">hvind</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;HaploidVariantIndex&gt;
  500 diploid variants on 22 chromosome(s)
  MAF ranges from nan to nan
  0 annotation(s)
         vid  chrom zero_allele one_allele  af hcopy  pos_bp  pos_cM
variant
0.0        0      0           A          G NaN     0     NaN     NaN
0.1        0      0           A          G NaN     1     NaN     NaN
1.0        1      0           A          G NaN     0     NaN     NaN
1.1        1      0           A          G NaN     1     NaN     NaN
2.0        2      0           A          G NaN     0     NaN     NaN
...      ...    ...         ...        ...  ..   ...     ...     ...
497.1    497     21           A          G NaN     1     NaN     NaN
498.0    498     21           A          G NaN     0     NaN     NaN
498.1    498     21           A          G NaN     1     NaN     NaN
499.0    499     21           A          G NaN     0     NaN     NaN
499.1    499     21           A          G NaN     1     NaN     NaN

[1000 rows x 8 columns]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hvind</span><span class="o">.</span><span class="n">to_diploid</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;DiploidVariantIndex&gt;
  500 diploid variants on 22 chromosome(s)
  MAF ranges from nan to nan
  0 annotation(s)
         vid  chrom zero_allele one_allele  af hcopy  pos_bp  pos_cM
variant
0.d        0      0           A          G NaN     d     NaN     NaN
1.d        1      0           A          G NaN     d     NaN     NaN
2.d        2      0           A          G NaN     d     NaN     NaN
3.d        3      0           A          G NaN     d     NaN     NaN
4.d        4      0           A          G NaN     d     NaN     NaN
...      ...    ...         ...        ...  ..   ...     ...     ...
495.d    495     21           A          G NaN     d     NaN     NaN
496.d    496     21           A          G NaN     d     NaN     NaN
497.d    497     21           A          G NaN     d     NaN     NaN
498.d    498     21           A          G NaN     d     NaN     NaN
499.d    499     21           A          G NaN     d     NaN     NaN

[500 rows x 8 columns]
</pre></div></div>
</div>
</section>
</section>
<section id="Sample-indexing">
<h2>Sample indexing<a class="headerlink" href="#Sample-indexing" title="Permalink to this heading"></a></h2>
<p>We reference specific individuals using the <code class="docutils literal notranslate"><span class="pre">xft.index.SampleIndex</span></code> class which includes three individual-level pieces of information:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">iid</span></code>, a vector of individual IDs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fid</span></code>, a vector of family IDs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sex</span></code>, a vector of biological sexes, with 0 and 1 encoding female and male respectively,</p></li>
</ul>
<p>as well as the <code class="docutils literal notranslate"><span class="pre">generation</span></code> of the sample. Sample indexes are always specific to a single generation.</p>
<p>There are several ways to construct a <code class="docutils literal notranslate"><span class="pre">SampleIndex</span></code>, all of which will automatically construct unique identifiers:</p>
<section id="Constructing-a-generic-sample-index">
<h3>Constructing a generic sample index<a class="headerlink" href="#Constructing-a-generic-sample-index" title="Permalink to this heading"></a></h3>
<p>If all we want is a sample of arbitrarily-named, unrelated, sex-balanced individuals, we can simply provide the number of individuals <code class="docutils literal notranslate"><span class="pre">n</span></code> and the <code class="docutils literal notranslate"><span class="pre">generation</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">SampleIndex</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">generation</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;SampleIndex&gt;
  Generation 1
  5 indviduals from 5 families
  3 biological females
  2 biological males
            iid  fid  sex
sample
1..1_0.1_0  1_0  1_0    0
1..1_1.1_1  1_1  1_1    1
1..1_2.1_2  1_2  1_2    0
1..1_3.1_3  1_3  1_3    1
1..1_4.1_4  1_4  1_4    0
</pre></div></div>
</div>
</section>
<section id="Constructing-a-sample-index-for-specific-iids-and-fids">
<h3>Constructing a sample index for specific iids and fids<a class="headerlink" href="#Constructing-a-sample-index-for-specific-iids-and-fids" title="Permalink to this heading"></a></h3>
<p>We can alternatively provide specific iids, fids, and sexes. Here we create a sample index comprised of two families: three sisters in the first and two brothers in the second:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sind</span> <span class="o">=</span> <span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">SampleIndex</span><span class="p">(</span><span class="n">iid</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;0_sister1&#39;</span><span class="p">,</span><span class="s1">&#39;0_sister2&#39;</span><span class="p">,</span><span class="s1">&#39;0_sister3&#39;</span><span class="p">,</span><span class="s1">&#39;0_brother1&#39;</span><span class="p">,</span> <span class="s1">&#39;0_brother2&#39;</span><span class="p">],</span>
                             <span class="n">fid</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span>
                             <span class="n">sex</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">generation</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">sind</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;SampleIndex&gt;
  Generation 0
  5 indviduals from 2 families
  3 biological females
  2 biological males
                        iid  fid  sex
sample
0..0_sister1.0    0_sister1    0    0
0..0_sister2.0    0_sister2    0    0
0..0_sister3.0    0_sister3    0    0
0..0_brother1.1  0_brother1    1    1
0..0_brother2.1  0_brother2    1    1
</pre></div></div>
</div>
</section>
<section id="Constructing-a-sample-index-with-a-DataFrame">
<h3>Constructing a sample index with a DataFrame<a class="headerlink" href="#Constructing-a-sample-index-with-a-DataFrame" title="Permalink to this heading"></a></h3>
<p>Finally, we can simply provide a Pandas DataFrame using the <code class="docutils literal notranslate"><span class="pre">frame</span></code> argument. In this case we also need to specify the generation.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sind</span><span class="o">.</span><span class="n">frame</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>iid</th>
      <th>fid</th>
      <th>sex</th>
    </tr>
    <tr>
      <th>sample</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0..0_sister1.0</th>
      <td>0_sister1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>0..0_sister2.0</th>
      <td>0_sister2</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>0..0_sister3.0</th>
      <td>0_sister3</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>0..0_brother1.1</th>
      <td>0_brother1</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>0..0_brother2.1</th>
      <td>0_brother2</td>
      <td>1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">xft</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">SampleIndex</span><span class="p">(</span><span class="n">frame</span><span class="o">=</span><span class="n">sind</span><span class="o">.</span><span class="n">frame</span><span class="p">,</span> <span class="n">generation</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;SampleIndex&gt;
  Generation 1
  5 indviduals from 2 families
  3 biological females
  2 biological males
                        iid  fid  sex
sample
1..0_sister1.0    0_sister1    0    0
1..0_sister2.0    0_sister2    0    0
1..0_sister3.0    0_sister3    0    0
1..0_brother1.1  0_brother1    1    1
1..0_brother2.1  0_brother2    1    1
</pre></div></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="nutsandbolts.html" class="btn btn-neutral float-left" title="Nuts and bolts" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="struct.html" class="btn btn-neutral float-right" title="Data structures" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Richard Border.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>